### ISSUE TITLE
Significant hangs on some Epichrome apps

### ISSUE THREAD

**Issue by ylluminate** [dmarmor/epichrome#182]
I'm having a very peculiar issue that's driving me quite crazy. I have a few Epichrome apps that seem to want to hang for long periods of time (15-30 minutes) sometimes. The bad thing is that I can't really reproduce this in a reliable fashion and I can't seem to get Chrome proper to do it. I'm being lead to believe that it's something specific to Epichrome instances, but I can't really isolate the problem.

I've got a hard hang of an Epichrome app instance right now and sampled it for review here in case you can see any particular problem point: 
https://cl.ly/a3cef4102370/2019-03-19%20(18-45)%20-%20Sample%20of%20EpiAppName.txt

**Comment by dmarmor** [comment]
Hey @ylluminate, I just took a quick look at the sample, but couldn't immediately see anything useful in it. Best I could say right now is keep looking for any patterns. If you can figure out how to reproduce it even semi-reliably then I can set up the same situation on my system and try to get to the bottom of it.

For background, what version of macOS are you running, and do the apps that hang have anything in common in terms of what sites they point to?

**Comment by ylluminate** [comment]
Yeah, that's my problem. I can't seem to isolate anything and it's MOSTLY happening with Epichrome apps. 10.14.4 build 18E220a. 

I'm wondering if it's due to how Chrome instantiates so many processes.  Hrm.

Otherwise I'm having a devil of a time tracking down any commonalities. Just keeps happening and happening and sometimes everything is fine. Not sure what I can look for that would be blocking or halting the processes.

**Comment by ylluminate** [comment]
Hrm, keep having these hangs. I'm wondering if there's a way to grok these samples to in order to get to the bottom of where the hang's happening in the process(es):
https://cl.ly/5a850e237ddc/2019-03-24%20(19-56)%20-%20epichrome%20app

This hang happened when I hit cmd-q to quit the app!  O.o

**Comment by ylluminate** [comment]
One of the craziest things that I've seen so far here with this in terms of reproducibility is that the only guarantee I have seems to be that when I make a new Epichrome app or sign into one for the first time I can count on a hang about 50% or more of the time. I just made a new Epichrome app and upon attempting to simply paste a username from a site into a field, the entire thing has now hung and I can count on the hang lasting roughly 15-25 minutes before I get to use the app.  Really insane.

**Comment by ylluminate** [comment]
One thing I may have found is that pasting has been, for some reason, consistently hanging a newly create Epichrome app for me when pasting in a password from 1Password 7.app and using Pastebot.

**Comment by ylluminate** [comment]
I believe I found the issue. Somehow it appears that Chrome was set to be my default Mail app. I did not set this (to my knowledge) and so subsequently it appears that various invocations on the sharing extension / subsystem of macOS was getting really discombobulated with Chrome and probably hung for such long times since I have "many" Chrome apps via Epichrome. It has been a really strange issue, but since discovering my default mail app was changed a couple days ago and setting it to my preferred mail app (Canary), I have seen no more hangs with any Epichrome Chrome instances...

**Comment by ylluminate** [comment]
I have found another odd issue, that I thought I should ask you about @dmarmor.  Does Chrome itself use `lsof` for things?  For example, I found it going nuts here this:
![lsof](https://cl.ly/cdc62d4f7025/Image%202019-04-24%20at%206.37.00%20PM.png)

And then I noticed that I was down to under 100mb free RAM.  I then quit a few Epichrome apps which freed up about 14gb of RAM. Suddenly this `lsof` process went away along with the hung `Google Software Update`,  and things started behaving properly again...

Any thoughts on this overall?

**Comment by ylluminate** [comment]
So @dmarmor I'm still quite losing my mind here. Things were better, but now they're a little... different maybe?  I'm still getting serious hangs and they are not quite the same, but I'm getting rather upset.  The target of this upsettedness is Google and Chromium... It seems that you can't reach a human being that gives a flipping flying crap atm to help isolate and debug the issue(s) further.

It got me thinking, I really think we need to consider an Epichrome Microsoft Edge Epichrome fork for the macOS version of Edge when it's released soon here: https://www.microsoftedgeinsider.com/en-us/download

At least perhaps we can have a better contact with Microsoft that can actually help us.  One thing I'll say, as much as I HATE MicroShaft, they will at least talk to you on the dev side.

**Comment by dmarmor** [comment]
Hi @ylluminate, so sorry for the very slow reply. Work and life have still been keeping me entirely away from Epichrome lately.

Regarding the lsof problem: I suspect that must be a result of running an older version of Epichrome. The evilness of lsof was brought up in issue #155 and I removed it from Epichrome in last September's release. So if you're still seeing it, something must still be running an old Epichrome engine.

As for the other weird hanging behavior, has any of it been resolved in later updates of Chrome? I haven't seen any of the behavior you describe on my system, and can't figure out what would be causing it.

Finally, regarding MS Edge--if I ever get the time (which seems sadly far-off), I would love to do a Firefox version of Epichrome, which would have a few advantages over Chrome. I'm not sure Edge would be a top priority for me, as I feel it would likely have similar problems to Chrome in terms of being a closed system likely to change mercurially without notice, and break Epichrome. But I'm open to being convinced otherwise!

**Comment by ylluminate** [comment]
Hey @dmarmor I'm sorry, I should have updated this ticket.  I found out from the developer of App Tamer that that is the source of the `lsof` trouble.  He uses it and cannot figure out an alternative to it that would accomplish his goal.  I tried to help him to research options a little, but ran into a dead end as well.

I am, though, still going insane with hangs.  I can only figure it's Google's fault at this point since it is such a profound resource hog.

So yes, I'm thinking Firefox would probably be better as well.  I've been thinking Brave would also be better (a better choice than the new Edge) since it at least reduces the footprint a little. 

Ideally we could get an SSB working that would use SHARED libraries... But the problem is that Chrom[e|ium] extensions are just so darn useful.  Maybe Google is going to kill this usefulness in the coming updates, but I don't have a crystal ball.  I'm not sure what to do.  I have 128GB RAM and having several instances of Epichrome Chrome apps open still brings my system to its knees...

BUT for heaven's sake, Epichrome is amazing in terms of keeping things organized and mental sanity.  I really think using a shared "Epichrome" library instance would be the ideal solution, but even Electron apps suffer from this same issue.  No one really seems to care about this horrible void in the current browser ecosystem!

**Comment by dmarmor** [comment]
Yeah, the resource-management of browsers seems to be pretty bad across the board, and with the current architecture there's of course no way for Epichrome to have a shared library since all that ends up running are instances of Chrome.

OK to close this issue given that the problem appears to be Chrome's resource hogginess?

**Comment by ylluminate** [comment]
Yeah, I've gone ahead and closed it.  By the way, do you know of any way to reach Google Chrome devs about hanging issues?  Sometimes when I have these hanging issues (even with 60GB of free RAM), Epichrome instances seem to be the primary apps that are hung.  I'd like to figure out how to troubleshoot this and find out why, where, how, etc. these things are happening and it seems that Chrome devs might be the only ones who can help.

**Comment by dmarmor** [comment]
Unfortunately, I haven't had any luck reaching anyone at Chrome that I'm confident is actually responsive to any feedback. Google in general feels very insulated from user feedback (e.g. I've been begging them for years to get colored stars working in the mobile Gmail app, with no response whatsoever).

If you do figure anything out or find any evidence that the problem is Epichrome-related in a way I could do anything to fix, please let me know.

**Comment by ylluminate** [comment]
Thanks. Yeah, I don't know if there's anything that can be done for sure.  I wrote a script to kill all Chrome (helper) instances and parents in order to do a nuclear reset when Chrome starts to go sideways on me.  That's the only thing I've found so far that keeps my system running when I have a few to several instances of Epichrome apps going and things start to get wonky.  

FYI, you might want to keep an eye on this or add support (I'm sure it's going to rub someone the wrong way): https://groups.google.com/a/chromium.org/forum/#!topic/chromium-discuss/Ab0odDAv4JE

**Comment by dmarmor** [comment]
Oops! Check issue #185--I just also wrote a killepichrome script! Oh well, whichever one works better for you I guess.

Thanks for the forum link. I'll add my voice.

**Comment by dmarmor** [comment]
Hey, I finally had a chance to go to the Google forum, and I think Jakob Kummerow actually seemed to know what he was talking about (despite being kind of snippy about it). Assuming he really does understand the Chrome architecture (he certainly appears to know it better than I do), then apparently it is running on shared libraries already across Chrome processes (basically meaning open tabs). To the degree there's a problem with Epichrome specifically, it's likely because when multiple Epichrome apps are run, the system believes there are actually that many separate copies of Chrome on the system, each running independently, in which case they can't share any memory between them (they may, for all the system knows, be totally different versions of Chrome). So it's probably an insoluble problem from the Chrome developers' side.

All that said, there may be something else going on on your system. I mainly work on a 2016 MacBook Pro with only 16GB RAM, and I frequently have 4-5 Epichrome apps open, along with a main Chrome with as many as 30 tabs open, and I've never run into the kinds of hangs you've experienced.

I wouldn't know where to start to diagnose the problem, but it might not actually be Chrome's fault...

**Comment by ylluminate** [comment]
I guess the problem I'm having here is making them understand my question and thus why I've not replied to him (in addition to his "snippy" attitude). I believe there is a way to make Chrome recognize the same versions of itself running on the system and I also believe that this can be done with Electron apps.  The problem is that they're not creative (or driven) enough to consider solutions.

There could be, for example, a global setting/config file set for any instances of Chrom(e|ium) to access that indicates the version of libraries and what libraries can be shared betwixt the various instances.  They could get smarter about dependencies and libs.  I could probably throw out several other options, but these people don't have the gump to even think about or consider more deeply the situation.  Why, oh why on bloody earth can they not consider that due to the (sad) popularity of their browser that they should do things that would reduce the weight of their app's usage across the entire system, no matter what operating system is being used?

Apple is entirely sure it's Chrome's fault at this point and I've had arguments with them.  I just work differently - but they and so many ppl don't like "different" anymore when it conflicts with very well defined parameters that they consider acceptable.  <sigh>

**Comment by ylluminate** [comment]
So I'm back to some reliable hangs in Epichrome instances again here, but primarily only after a reboot and within the first 1-2 hours... Any idea of how to get meaningful debugging data?  I've captured samples, but it seems we must need more to track this down since Chromium folks didn't respond to [my previous request](https://groups.google.com/a/chromium.org/forum/#!searchin/chromium-discuss/Where$20do$20I$20go$20to$20start$20to$20resolve$20this$20issue$3F%7Csort:date/chromium-discuss/iDvMcHmwT_E/NhD2p1zFCQAJ) to get some input. 

I really have to figure this out. It's a debilitating issue and I don't know if it's bound to Chrom(e|ium) since some other apps hang too, but there should be no memory usage or process limit issue at the start since there's usually at least 70-90GB free at this point with relatively few process running.

**Comment by dmarmor** [comment]
Yeah, this would be driving me crazy too. I'm kind of at a loss how to even start to diagnose it, since I've never experienced anything like it on my system. I also have to admit I don't know how to read those sample documents. Are you able to decipher them to the level of seeing what function or library Chrome is in when the hangs occur? That would be a start, anyway.

I also assume these hangs are not confined to specific websites or a repeatable operation I'd be able to try to replicate on my system?

In absence of a redesign of Chrome that seems highly unlikely (Epichrome apps would definitely not qualify as a supported use case for them, I suspect), the best bet here is probably to try to figure out if there's any workaround I can do in Epichrome's design (though I'm kind of pessimistic about that too, given the architecture).

**Comment by ylluminate** [comment]
This is where I'm hitting walls. I cannot find any clear instructions or examples on how to really grok samples at this point.  Years ago I used to routinely share samples with developers who seemed to be able to use these to isolate problem areas of applications, however over the past 1-2 years it seems as though no one can read them anymore.  Unfortunately I never had the time to actually ask those devs who seemed to understand what was going on to explain in detail their process of understanding and analyzing these samples.

More recently when asking some devs, they don't seem to have the time or patience to explain anything and tell me to just read it carefully and seem to think I'll be able to isolate issues this way at least to some degree.

Furthermore, I've thought that perhaps a ktrace (dtrace) system call tracing during hangs might reveal some issues, but I've had trouble with that as well in some cases in terms of getting human-parsable-and-actionable info out of the output.  Some have said that kernel debugging / gdb could help, but is a very slow process.  I'm really shooting blanks here.  I don't understand why it has become seemingly impossible to isolate hang issues.

I've submitted about 15 reports to Apple so far and they have ALL been silent after asking me for some more details.  I'm well past my patience threshold here.  Something has to be done and documented for macOS debugging and troubleshooting for situations such as these.  We seem to be losing all of our meaningful tooling - or it's being castrated in the interest of trying to lock down the OS.

**Comment by dmarmor** [comment]
Yeah, I'm really sorry I don't understand those samples. I looked through thinking they would show timestamps on each call, I could use to see where the big delay was happening, but as best I could interpret the output, it doesn't appear to show that info.

Does the sample output appear dynamically? In other words, can you watch it outputting the tree as your app is running? I'm assuming not, or I'm guessing you'd have already done this, but I wondered if it were possible for you to watch the output and see where it stops when you hit a hang.

**Comment by ylluminate** [comment]
Hmm, so someone recommended that I start using `spindump` to gather information.  They had an observation that involved a number of `sh` instantiations from Chrome. He then makes the remark: "Each process is around for only 9 samples? Are you forkbombing yourself?"  Two Chrome examples in here are:

```
Process:         sh [30313]
UUID:            E302FB8D-F27D-3532-9EA9-76ADD4DC07B0
Path:            /bin/sh
Architecture:    x86_64
Parent:          Google Chrome [30253]
Responsible:     Google Chrome [30253]
UID:             501
Footprint:       936 KB
Start time:      2019-10-01 22:25:42 -0400
End time:        2019-10-01 22:25:52 -0400
Num samples:     567 (1-567)
CPU Time:        0.031s (84.9M cycles, 31.2M instructions, 2.72c/i)

  Thread 0x18ef8            DispatchQueue 140736228080616                       567 samples (1-567)       priority 31 (base 31)     cpu time 0.031s (84.9M cycles, 31.2M instructions, 2.72c/i)
  567  start + 1 (libdyld.dylib + 91093) [0x7fff7e82a3d5]
    567  ??? (sh + 10774) [0x104010a16]
      567  ??? (sh + 116850) [0x10402a872]
        567  ??? (sh + 117790) [0x10402ac1e]
          567  ??? (sh + 124995) [0x10402c843]
            567  ??? (sh + 127937) [0x10402d3c1]
              567  ??? (sh + 136144) [0x10402f3d0]
                567  ??? (sh + 131821) [0x10402e2ed]
                  567  ??? (sh + 124588) [0x10402c6ac]
                    567  ??? (sh + 123826) [0x10402c3b2]
                      567  ??? (sh + 117790) [0x10402ac1e]
                        567  ??? (sh + 123826) [0x10402c3b2]
                          567  ??? (sh + 117790) [0x10402ac1e]
                            567  ??? (sh + 123826) [0x10402c3b2]
                              567  ??? (sh + 117790) [0x10402ac1e]
                                567  ??? (sh + 123826) [0x10402c3b2]
                                  567  ??? (sh + 117790) [0x10402ac1e]
                                    567  ??? (sh + 123826) [0x10402c3b2]
                                      567  ??? (sh + 117790) [0x10402ac1e]
                                        567  ??? (sh + 123826) [0x10402c3b2]
                                          567  ??? (sh + 117790) [0x10402ac1e]
                                            567  ??? (sh + 123826) [0x10402c3b2]
                                              567  ??? (sh + 117790) [0x10402ac1e]
                                                567  ??? (sh + 124129) [0x10402c4e1]
                                                  564  ??? (sh + 135071) [0x10402ef9f]
                                                    564  ??? (sh + 117790) [0x10402ac1e]
                                                      563  ??? (sh + 128492) [0x10402d5ec]
                                                        563  ??? (sh + 156211) [0x104034233]
                                                          563  __wait4 + 10 (libsystem_kernel.dylib + 25430) [0x7fff7e964356]
                                                           *563  ??? (kernel + 7179408) [0xffffff80008d8c90] (blocked by wait4 on any child process)
                                                      1    ??? (sh + 128341) [0x10402d555]
                                                        1    ??? (sh + 113692) [0x104029c1c]
                                                         *1    hndl_alltraps + 223 (kernel + 1423951) [0xffffff800035ba4f]
                                                           *1    user_trap + 410 (kernel + 2997946) [0xffffff80004dbeba]
                                                             *1    ??? (smbfs + 2023035010) [0xffffff8000198082] (running)
                                                  3    ??? (sh + 135142) [0x10402efe6]
                                                    3    ??? (sh + 117790) [0x10402ac1e]
                                                      2    ??? (sh + 128492) [0x10402d5ec]
                                                        2    ??? (sh + 156211) [0x104034233]
                                                          2    __wait4 + 10 (libsystem_kernel.dylib + 25430) [0x7fff7e964356]
                                                           *2    ??? (kernel + 7179408) [0xffffff80008d8c90] (blocked by wait4 on any child process)
                                                      1    ??? (sh + 128232) [0x10402d4e8]
                                                        1    ??? (sh + 154205) [0x104033a5d]
                                                          1    __fork + 11 (libsystem_kernel.dylib + 31431) [0x7fff7e965ac7]
                                                           *1    hndl_unix_scall64 + 22 (kernel + 1426182) [0xffffff800035c306]
                                                             *1    unix_syscall64 + 637 (kernel + 8103405) [0xffffff80009ba5ed]
                                                               *1    fork + 57 (kernel + 7089097) [0xffffff80008c2bc9]
                                                                 *1    fork1 + 630 (kernel + 7084278) [0xffffff80008c18f6]
                                                                   *1    cloneproc + 142 (kernel + 7087470) [0xffffff80008c256e]
                                                                     *1    fork_create_child + 67 (kernel + 7087667) [0xffffff80008c2633]
                                                                       *1    task_create_internal + 165 (kernel + 1943317) [0xffffff80003da715]
                                                                         *1    vm_map_fork + 2174 (kernel + 2441102) [0xffffff8000453f8e]
                                                                           *1    vm_map_store_entry_link + 189 (kernel + 2463021) [0xffffff800045952d]
                                                                             *1    update_holes_on_entry_creation + 76 (kernel + 2469660) [0xffffff800045af1c] (running)

  Binary Images:
           0x10400e000 -        0x10408ffff  sh (106.220.2)                            <E302FB8D-F27D-3532-9EA9-76ADD4DC07B0>  /bin/sh
        0x7fff7e814000 -     0x7fff7e840fff  libdyld.dylib (655.1.1)                   <002418CC-AD11-3D10-865B-015591D24E6C>  /usr/lib/system/libdyld.dylib
        0x7fff7e95e000 -     0x7fff7e986fff  libsystem_kernel.dylib (4903.271.2)       <EA204E3C-870B-30DD-B4AF-D1BB66420D14>  /usr/lib/system/libsystem_kernel.dylib
   *0xffffff7f87847000 - 0xffffff7f878a7fff  com.apple.filesystems.smbfs 3.3.2 (3.3.2) <A2F3A4BF-DA29-3079-8B50-80CB22816DAF>  /System/Library/Extensions/smbfs.kext/Contents/MacOS/smbfs
   *0xffffff8000200000 - 0xffffff8000bfffff  kernel (4903.271.2)                       <C41337A1-0EC3-3896-A954-A1F85E849D53>  /System/Library/Kernels/kernel


Process:         sh [30536]
UUID:            E302FB8D-F27D-3532-9EA9-76ADD4DC07B0
Path:            /bin/sh
Architecture:    x86_64
Parent:          Google Chrome [30397]
Responsible:     Google Chrome [30397]
UID:             501
Footprint:       900 KB
Start time:      2019-10-01 22:25:42 -0400
End time:        2019-10-01 22:25:52 -0400
Num samples:     567 (1-567)
CPU Time:        0.024s (73.6M cycles, 25.6M instructions, 2.88c/i)

  Thread 0x19310            DispatchQueue 140736228080616                       567 samples (1-567)       priority 31 (base 31)     cpu time 0.024s (73.6M cycles, 25.6M instructions, 2.88c/i)
  567  start + 1 (libdyld.dylib + 91093) [0x7fff7e82a3d5]
    567  ??? (sh + 10774) [0x10017fa16]
      567  ??? (sh + 116850) [0x100199872]
        567  ??? (sh + 117790) [0x100199c1e]
          567  ??? (sh + 124995) [0x10019b843]
            567  ??? (sh + 127937) [0x10019c3c1]
              567  ??? (sh + 136144) [0x10019e3d0]
                567  ??? (sh + 131821) [0x10019d2ed]
                  567  ??? (sh + 124588) [0x10019b6ac]
                    567  ??? (sh + 123826) [0x10019b3b2]
                      567  ??? (sh + 117790) [0x100199c1e]
                        567  ??? (sh + 123826) [0x10019b3b2]
                          567  ??? (sh + 117790) [0x100199c1e]
                            567  ??? (sh + 123826) [0x10019b3b2]
                              567  ??? (sh + 117790) [0x100199c1e]
                                567  ??? (sh + 123826) [0x10019b3b2]
                                  567  ??? (sh + 117790) [0x100199c1e]
                                    567  ??? (sh + 123826) [0x10019b3b2]
                                      567  ??? (sh + 117790) [0x100199c1e]
                                        567  ??? (sh + 123826) [0x10019b3b2]
                                          567  ??? (sh + 117790) [0x100199c1e]
                                            567  ??? (sh + 123826) [0x10019b3b2]
                                              567  ??? (sh + 117790) [0x100199c1e]
                                                567  ??? (sh + 124129) [0x10019b4e1]
                                                  564  ??? (sh + 135071) [0x10019df9f]
                                                    564  ??? (sh + 117790) [0x100199c1e]
                                                      564  ??? (sh + 128492) [0x10019c5ec]
                                                        564  ??? (sh + 156211) [0x1001a3233]
                                                          564  __wait4 + 10 (libsystem_kernel.dylib + 25430) [0x7fff7e964356]
                                                           *564  ??? (kernel + 7179408) [0xffffff80008d8c90] (blocked by wait4 on any child process)
                                                  3    ??? (sh + 135142) [0x10019dfe6]
                                                    3    ??? (sh + 117790) [0x100199c1e]
                                                      2    ??? (sh + 128492) [0x10019c5ec]
                                                        2    ??? (sh + 156211) [0x1001a3233]
                                                          2    __wait4 + 10 (libsystem_kernel.dylib + 25430) [0x7fff7e964356]
                                                           *2    ??? (kernel + 7179408) [0xffffff80008d8c90] (blocked by wait4 on any child process)
                                                      1    ??? (sh + 128232) [0x10019c4e8]
                                                        1    ??? (sh + 154205) [0x1001a2a5d]
                                                          1    __fork + 11 (libsystem_kernel.dylib + 31431) [0x7fff7e965ac7]
                                                           *1    hndl_unix_scall64 + 22 (kernel + 1426182) [0xffffff800035c306]
                                                             *1    unix_syscall64 + 637 (kernel + 8103405) [0xffffff80009ba5ed]
                                                               *1    fork + 57 (kernel + 7089097) [0xffffff80008c2bc9]
                                                                 *1    fork1 + 630 (kernel + 7084278) [0xffffff80008c18f6]
                                                                   *1    cloneproc + 142 (kernel + 7087470) [0xffffff80008c256e]
                                                                     *1    fork_create_child + 67 (kernel + 7087667) [0xffffff80008c2633]
                                                                       *1    task_create_internal + 165 (kernel + 1943317) [0xffffff80003da715]
                                                                         *1    vm_map_fork + 2131 (kernel + 2441059) [0xffffff8000453f63]
                                                                           *1    ??? (kernel + 2489726) [0xffffff800045fd7e] (running)

  Binary Images:
           0x10017d000 -        0x1001fefff  sh (106.220.2)                      <E302FB8D-F27D-3532-9EA9-76ADD4DC07B0>  /bin/sh
        0x7fff7e814000 -     0x7fff7e840fff  libdyld.dylib (655.1.1)             <002418CC-AD11-3D10-865B-015591D24E6C>  /usr/lib/system/libdyld.dylib
        0x7fff7e95e000 -     0x7fff7e986fff  libsystem_kernel.dylib (4903.271.2) <EA204E3C-870B-30DD-B4AF-D1BB66420D14>  /usr/lib/system/libsystem_kernel.dylib
   *0xffffff8000200000 - 0xffffff8000bfffff  kernel (4903.271.2)                 <C41337A1-0EC3-3896-A954-A1F85E849D53>  /System/Library/Kernels/kernel
```

**Comment by dmarmor** [comment]
Hmm. This does not look like anything that should be related to Epichrome. If I'm reading this right, it looks like the first call to sh is from libdyld.dylib. The only way I can figure that could be related to Epichrome is if it has something to do with the Epichrome Helper extension, which talks with a python-based native messaging host. That would possibly be launched through an sh call in libdyld.dylib.

But I cannot think of any way that could be making all those nested calls to sh. Literally all that host does is wait for a link request, then send it to the default browser...
